custom/enikshay/tasks.py

```diff
@@ -0,0 +1,295 @@
+
+            if only enikshay source cases
+                filter by '(closed and closure_reason == HISTORICAL_CLOSURE_REASON) or open'
+                consider the case with latest modified after above filter
+
+            if mix of enikshay and non-enikshay source cases
+                ignore non-enikshay and apply above enikshay only condition
+        2. Check if 'adherence_value' of most relevent case is one of DOSE_TAKEN_INDICATORS
+        """
+        def is_dose_taken(cases):
+            # runs above discribed calculation and returns whether a dose is taken or not
+            sources = set(map(lambda x: x["adherence_source"], cases))
+            if 'enikshay' not in sources:
+                valid_cases = cases
+            else:
+                valid_cases = filter(
+                    lambda case: (
+                        case['adherence_source'] is 'enikshay' and
+                        (not case['closed'] or (case['closed'] and
+                         case['adherence_closure_reason'] == HISTORICAL_CLOSURE_REASON))
+                    ),
+                    cases
+                )
+            if valid_cases:
+                by_modified_on = sorted(valid_cases, key=lambda case: case['modified_on'])
+                return by_modified_on[-1]['adherence_value'] in DOSE_TAKEN_INDICATORS
+            else:
+                return False
+
+        # index by 'adherence_date'
+        cases_by_date = defaultdict(list)
+        for case in adherence_cases:
+            adherence_date = parse_datetime(case['adherence_date']).date()
+            cases_by_date[adherence_date].append(case)
+
+        # calculate whether adherence is taken on each day
+        dose_taken_by_date = defaultdict(bool)
+        for d, cases in cases_by_date.iteritems():
+            dose_taken_by_date[d] = is_dose_taken(cases)
+
+        return dose_taken_by_date
+
+    def count_doses_taken(self, adherence_cases, lte=None, gte=None):
+        """
+        Args:
+            adherence_cases: list of 'adherence_cases' with its 'adherence_value' as one of DOSE_KNOWN_INDICATORS
+
+        Returns:
+            total count of adherence_cases excluding duplicates on a given day. If there are
+            two adherence_cases on one day at different time, it will be counted as one
+        """
+
+        if not self._cache_dose_taken_by_date and self._cache_dose_taken_by_date != {}:
+            # we can't memoize this as the arg is list.
+            dose_taken_by_date = self.calculate_doses_taken_by_day(adherence_cases)
+        else:
+            dose_taken_by_date = self._cache_dose_taken_by_date
+
+        if bool(lte) != bool(lte):
+            raise Exception("Both of lte and gte should be specified or niether of them")
+
+        if not lte:
+            return dose_taken_by_date.values().count(True)
+        else:
+            # any efficient way to do this - numpy, python bisect?
```

Comment:
This doesn't look like it will be a bottleneck to me. Filtering this list is just `O(n)`.